{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import sqlite3\n",
    "# import gffutils\n",
    "import time\n",
    "# import pybedtools\n",
    "import os\n",
    "from IPython import get_ipython\n",
    "from multiprocessing import Pool\n",
    "import ctypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_file_name = \"gencode.vM36.annotation.gtf\"\n",
    "# gtf_file_name = \"/home/U210050044/data/DH607-project/gencode.vM36.annotation.gtf\"\n",
    "# sql_db_name = \"db-mouse.sqlite3\"\n",
    "sql_db_name = \"file:mouse?mode=memory&cache=shared\"\n",
    "results_file_name = \"results_mouse.txt\"\n",
    "os.system(\"g++ -std=c++11 -shared -fPIC -o gtf_to_sql.so gtf_to_sql.cpp -lsqlite3 -pthread\")\n",
    "lib = ctypes.CDLL(\"./gtf_to_sql.so\")\n",
    "# Define the function signature\n",
    "lib.run_gtf_to_sql.argtypes = [\n",
    "    ctypes.c_char_p,  # db_name\n",
    "    ctypes.c_char_p,  # table_name\n",
    "    ctypes.c_char_p,  # input_file\n",
    "    ctypes.c_int,     # num_threads\n",
    "]\n",
    "OUTPUT_MODE = \"print\"\n",
    "# OUTPUT_MODE = \"file\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "if OUTPUT_MODE == \"file\":\n",
    "    results_file = open(results_file_name,\"w\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished reading into pyranges object in time 85.17597937583923\n"
     ]
    }
   ],
   "source": [
    "start_time = time.time()\n",
    "mouse_gr = pr.read_gtf(gtf_file_name)\n",
    "end_time = time.time()\n",
    "if OUTPUT_MODE == \"file\":\n",
    "    results_file.write(f\"Finished reading into pyranges object in time {end_time-start_time}\\n\")\n",
    "else:\n",
    "    print(f\"Finished reading into pyranges object in time {end_time-start_time}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time taken to read data: 16ms\n",
      "Time taken to read data: 12ms\n",
      "Time taken to read data: 12ms\n",
      "Time taken to read data: 11ms\n",
      "Time taken to read data: 12ms\n",
      "Time taken to read data: 17ms\n",
      "Time taken to read data: 22ms\n",
      "Time taken to read data: 27ms\n",
      "Time taken to process data: 521ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 532ms\n",
      "Time taken to process data: 513ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 589ms\n",
      "Time taken to process data: 515ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 586ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 584ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 701ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 450ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 452ms\n",
      "Time taken to process data: 474ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 482ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 472ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 495ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 454ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 459ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 454ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 466ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 486ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 478ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 469ms\n",
      "Time taken to move data: 0ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 462ms\n",
      "Time taken to process data: 451ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 450ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 512ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 476ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 476ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 495ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 465ms\n",
      "Time taken to process data: 480ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 772ms\n",
      "Time taken to read data: 6ms\n",
      "Time taken to process data: 566ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 486ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 484ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 439ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 487ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 504ms\n",
      "Time taken to process data: 536ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 471ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 480ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 492ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 528ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 476ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 770ms\n",
      "Time taken to read data: 6ms\n",
      "Time taken to process data: 506ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 529ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 536ms\n",
      "Time taken to read data: 6ms\n",
      "Time taken to process data: 471ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 475ms\n",
      "Time taken to read data: 5ms\n",
      "Time taken to process data: 529ms\n",
      "Time taken to read data: 6ms\n",
      "Time taken to process data: 587ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 497ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 502ms\n",
      "Time taken to process data: 767ms\n",
      "Time taken to process data: 502ms\n",
      "Time taken to process data: 482ms\n",
      "Time taken to process data: 540ms\n",
      "Time taken to read data: 7ms\n",
      "Time taken to read data: 8ms\n",
      "Time taken to read data: 5ms\n",
      "Time taken to read data: 5ms\n",
      "Time taken to read data: 6ms\n",
      "Time taken to process data: 520ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 557ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 495ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 496ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 516ms\n",
      "Time taken to process data: 513ms\n",
      "Time taken to read data: 5ms\n",
      "Time taken to read data: 4ms\n",
      "Processed 100000 lines\n",
      "Time taken to process data: 530ms\n",
      "Time taken to read data: 5ms\n",
      "Time taken to process data: 479ms\n",
      "Time taken to read data: 4ms\n",
      "Processed 100000 lines\n",
      "Time taken to process data: 549ms\n",
      "Time taken to read data: 4ms\n",
      "Processed 100000 lines\n",
      "Time taken to process data: 503ms\n",
      "Time taken to read data: 4ms\n",
      "Processed 100000 lines\n",
      "Time taken to process data: 851ms\n",
      "Time taken to read data: 6ms\n",
      "Time taken to process data: 506ms\n",
      "Time taken to read data: 4ms\n",
      "Processed 100000 lines\n",
      "Time taken to process data: 482ms\n",
      "Time taken to process data: 484ms\n",
      "Time taken to read data: 4ms\n",
      "Processed 100000 lines\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 508ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 485ms\n",
      "Time taken to read data: 5ms\n",
      "Time taken to process data: 485ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 503ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 459ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 483ms\n",
      "Time taken to read data: 6ms\n",
      "Time taken to process data: 517ms\n",
      "Time taken to read data: 5ms\n",
      "Time taken to process data: 485ms\n",
      "Time taken to read data: 5ms\n",
      "Time taken to process data: 508ms\n",
      "Time taken to read data: 5ms\n",
      "Time taken to process data: 523ms\n",
      "Time taken to process data: 853ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to read data: 6ms\n",
      "Time taken to process data: 486ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 503ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 502ms\n",
      "Time taken to process data: 494ms\n",
      "Time taken to process data: 565ms\n",
      "Time taken to process data: 509ms\n",
      "Time taken to process data: 516ms\n",
      "Time taken to process data: 487ms\n",
      "Time taken to read data: 5ms\n",
      "Time taken to read data: 5ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 671ms\n",
      "Time taken to read data: 7ms\n",
      "Time taken to read data: 6ms\n",
      "Time taken to read data: 7ms\n",
      "Time taken to read data: 7ms\n",
      "Processed 100000 lines\n",
      "Time taken to process data: 513ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 512ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 527ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 454ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 712ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 766ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 877ms\n",
      "Time taken to read data: 6ms\n",
      "Time taken to process data: 900ms\n",
      "Time taken to process data: 918ms\n",
      "Time taken to read data: 7ms\n",
      "Time taken to read data: 6ms\n",
      "Time taken to process data: 466ms\n",
      "Time taken to read data: 4ms\n",
      "Pushed 209995 lines to db in time 5842ms\n",
      "Time taken to process data: 478ms\n",
      "Time taken to move data: 0ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 469ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 505ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 489ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 467ms\n",
      "Time taken to read data: 5ms\n",
      "Time taken to process data: 481ms\n",
      "Time taken to process data: 452ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 762ms\n",
      "Time taken to read data: 5ms\n",
      "Time taken to process data: 446ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 463ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 885ms\n",
      "Time taken to process data: 878ms\n",
      "Time taken to read data: 6ms\n",
      "Time taken to read data: 6ms\n",
      "Time taken to process data: 450ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 453ms\n",
      "Time taken to process data: 466ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 446ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 492ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 469ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 472ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 480ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 505ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 507ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 853ms\n",
      "Time taken to process data: 546ms\n",
      "Time taken to read data: 6ms\n",
      "Time taken to process data: 873ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to read data: 6ms\n",
      "Time taken to process data: 456ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 458ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 483ms\n",
      "Time taken to read data: 3ms\n",
      "Processed 200000 lines\n",
      "Time taken to process data: 466ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 470ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 506ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 466ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 478ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 452ms\n",
      "Time taken to process data: 461ms\n",
      "Time taken to process data: 845ms\n",
      "Time taken to process data: 866ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 469ms\n",
      "Time taken to read data: 6ms\n",
      "Time taken to read data: 6ms\n",
      "Time taken to read data: 6ms\n",
      "Time taken to process data: 454ms\n",
      "Time taken to read data: 4ms\n",
      "Processed 200000 lines\n",
      "Time taken to process data: 472ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 491ms\n",
      "Time taken to read data: 4ms\n",
      "Processed 200000 lines\n",
      "Time taken to process data: 446ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 451ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 470ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 433ms\n",
      "Time taken to read data: 5ms\n",
      "Processed 200000 lines\n",
      "Time taken to process data: 535ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 439ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 869ms\n",
      "Time taken to process data: 844ms\n",
      "Time taken to read data: 5ms\n",
      "Time taken to read data: 5ms\n",
      "Processed 200000 lines\n",
      "Time taken to process data: 505ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 500ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 534ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 525ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 529ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 438ms\n",
      "Time taken to process data: 449ms\n",
      "Time taken to process data: 719ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to read data: 5ms\n",
      "Time taken to process data: 440ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 467ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 459ms\n",
      "Time taken to read data: 3ms\n",
      "Processed 200000 lines\n",
      "Time taken to process data: 446ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 440ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 427ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 446ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 425ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 435ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 444ms\n",
      "Time taken to process data: 448ms\n",
      "Time taken to process data: 442ms\n",
      "Time taken to process data: 771ms\n",
      "Time taken to process data: 424ms\n",
      "Time taken to process data: 447ms\n",
      "Time taken to process data: 434ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to read data: 5ms\n",
      "Time taken to process data: 438ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 405ms\n",
      "Time taken to process data: 412ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 420ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 425ms\n",
      "Time taken to process data: 422ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 405ms\n",
      "Time taken to read data: 4ms\n",
      "Processed 200000 lines\n",
      "Time taken to process data: 810ms\n",
      "Time taken to read data: 5ms\n",
      "Time taken to process data: 419ms\n",
      "Time taken to process data: 805ms\n",
      "Time taken to process data: 418ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to read data: 5ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 424ms\n",
      "Time taken to process data: 447ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 444ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 445ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 392ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 429ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 405ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 389ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 416ms\n",
      "Time taken to process data: 442ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 776ms\n",
      "Time taken to read data: 5ms\n",
      "Time taken to process data: 808ms\n",
      "Time taken to read data: 5ms\n",
      "Time taken to process data: 450ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 432ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 440ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 442ms\n",
      "Time taken to process data: 442ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 526ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to process data: 399ms\n",
      "Time taken to process data: 439ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 398ms\n",
      "Time taken to read data: 3ms\n",
      "Processed 300000 lines\n",
      "Time taken to process data: 440ms\n",
      "Time taken to process data: 435ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 419ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 819ms\n",
      "Time taken to read data: 6ms\n",
      "Time taken to process data: 831ms\n",
      "Time taken to read data: 5ms\n",
      "Time taken to process data: 425ms\n",
      "Time taken to read data: 3ms\n",
      "Processed 300000 lines\n",
      "Time taken to process data: 430ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 448ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 448ms\n",
      "Time taken to process data: 436ms\n",
      "Time taken to process data: 461ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 430ms\n",
      "Time taken to process data: 444ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to read data: 6ms\n",
      "Time taken to process data: 420ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 425ms\n",
      "Time taken to read data: 3ms\n",
      "Processed 300000 lines\n",
      "Time taken to process data: 426ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 469ms\n",
      "Time taken to read data: 6ms\n",
      "Processed 300000 lines\n",
      "Time taken to process data: 835ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 824ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 430ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 425ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 443ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 440ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 451ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 450ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 829ms\n",
      "Time taken to read data: 5ms\n",
      "Time taken to process data: 439ms\n",
      "Time taken to read data: 3ms\n",
      "Processed 300000 lines\n",
      "Time taken to process data: 430ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to process data: 829ms\n",
      "Time taken to process data: 433ms\n",
      "Time taken to process data: 438ms\n",
      "Time taken to process data: 433ms\n",
      "Time taken to process data: 439ms\n",
      "Time taken to process data: 424ms\n",
      "Time taken to process data: 457ms\n",
      "Time taken to process data: 526ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to read data: 3ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to read data: 5ms\n",
      "Time taken to read data: 4ms\n",
      "Time taken to read data: 5ms\n",
      "Time taken to read data: 0ms\n",
      "Time taken to process data: 0ms\n",
      "Time taken to read data: 0ms\n",
      "Time taken to process data: 0ms\n",
      "Time taken to process data: 414ms\n",
      "Time taken to process data: 436ms\n",
      "Time taken to read data: 0ms\n",
      "Time taken to process data: 0ms\n",
      "Time taken to read data: 0ms\n",
      "Time taken to process data: 0ms\n",
      "Time taken to process data: 451ms\n",
      "Time taken to process data: 443ms\n",
      "Time taken to read data: 0ms\n",
      "Time taken to process data: 0ms\n",
      "Time taken to read data: 0ms\n",
      "Time taken to process data: 0ms\n",
      "Time taken to process data: 448ms\n",
      "Time taken to process data: 564ms\n",
      "Time taken to read data: 0ms\n",
      "Time taken to process data: 0ms\n",
      "Pushed 810000 lines to db in time 13801ms\n",
      "Time taken to move data: 0ms\n",
      "Pushed 1458103 lines to db in time 22030ms\n",
      "Finished reading into sql db in time 47.81519842147827\n",
      "Time taken to create indices: 4400ms\n"
     ]
    }
   ],
   "source": [
    "# Call the function\n",
    "start_time = time.time()\n",
    "lib.run_gtf_to_sql(sql_db_name.encode(\"utf-8\"), \"mouse\".encode(\"utf-8\"), gtf_file_name.encode(\"utf-8\"), 8)\n",
    "end_time = time.time()\n",
    "if OUTPUT_MODE == \"file\":\n",
    "    results_file.write(f\"Finished writing into sql db in time {end_time-start_time}\\n\")\n",
    "else:\n",
    "    print(f\"Finished writing into sql db in time {end_time-start_time}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "conn = sqlite3.connect(sql_db_name)\n",
    "cur = conn.cursor()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "start_time = time.time()\n",
    "mouse_gr2 = pr.PyRanges(pd.read_sql(\"SELECT * FROM mouse\", conn))\n",
    "end_time = time.time()\n",
    "if OUTPUT_MODE == \"file\":\n",
    "    results_file.write(f\"Finished reading from sql to pyranges in time {end_time-start_time}\\n\")\n",
    "else:\n",
    "    print(f\"Finished reading from sql to pyranges in time {end_time-start_time}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Aggregate Queries: Counting transcripts, calculating exon lengths, and grouping features\n",
    "### Aggregate Query 1: Count the number of exons for each gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using pyranges\n",
    "time = get_ipython().run_line_magic(\"timeit\",'-o (mouse_gr[mouse_gr.Feature == \"exon\"].df.groupby(\"gene_id\").size().reset_index(name=\"exon_count\"))')\n",
    "if OUTPUT_MODE == \"file\":\n",
    "    results_file.write(f\"Time taken to count exons using pyranges is {time}\\n\")\n",
    "else:\n",
    "    print(f\"Time taken to count exons using pyranges is {time}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using sql\n",
    "time = get_ipython().run_line_magic(\"timeit\",'-o exon_counts_sql = pd.read_sql_query(\"SELECT gene_id, COUNT(*) as exon_count FROM mouse WHERE Feature = \\'exon\\' GROUP BY gene_id\", conn)')\n",
    "if OUTPUT_MODE == \"file\":\n",
    "    results_file.write(f\"Time taken to count exons using sql is {time}\\n\")\n",
    "else:\n",
    "    print(f\"Time taken to count exons using sql is {time}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "exon counts are identical\n"
     ]
    }
   ],
   "source": [
    "# check if exon_counts_pr and exon_counts_sql are identical\n",
    "exon_counts_pr = (mouse_gr[mouse_gr.Feature == \"exon\"].df.groupby(\"gene_id\").size().reset_index(name=\"exon_count\"))\n",
    "exon_counts_sql = pd.read_sql_query(\"SELECT gene_id, COUNT(*) as exon_count FROM mouse WHERE Feature = 'exon' GROUP BY gene_id\", conn)\n",
    "exon_counts_sql[\"exon_count\"] = exon_counts_sql[\"exon_count\"].astype(int)\n",
    "exon_counts_pr[\"exon_count\"] = exon_counts_pr[\"exon_count\"].astype(int)\n",
    "if exon_counts_sql.equals(exon_counts_pr):\n",
    "    result_text = \"exon counts are identical\"\n",
    "else:\n",
    "    result_text = \"exon counts are not identical\"\n",
    "if OUTPUT_MODE == \"file\":\n",
    "    results_file.write(f\"{result_text}\\n\")\n",
    "else:\n",
    "    print(result_text)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Aggregate Query 2: Calculate the total length of exons for each gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using pyranges\n",
    "time = get_ipython().run_line_magic(\"timeit\",'-o exon_lengths_pr = (mouse_gr[mouse_gr.Feature == \"exon\"].df.assign(length=lambda df: df[\"End\"] - df[\"Start\"]).groupby(\"gene_id\")[\"length\"].sum().reset_index(name=\"total_exon_length\"))')\n",
    "if OUTPUT_MODE == \"file\":\n",
    "    results_file.write(f\"Time taken to calculate total exon length using pyranges is {time}\\n\")\n",
    "else:\n",
    "    print(f\"Time taken to calculate total exon length using pyranges is {time}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# using sql\n",
    "time = get_ipython().run_line_magic(\"timeit\",'-o exon_lengths_sql = pd.read_sql_query(\"SELECT gene_id, SUM(End - Start) as total_exon_length FROM mouse WHERE Feature = \\'exon\\' GROUP BY gene_id\", conn)')\n",
    "if OUTPUT_MODE == \"file\":\n",
    "    results_file.write(f\"Time taken to calculate total exon length using sql is {time}\\n\")\n",
    "else:\n",
    "    print(f\"Time taken to calculate total exon length using sql is {time}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "exon lengths are identical\n"
     ]
    }
   ],
   "source": [
    "# check if exon_lengths_pr and exon_lengths_sql are identical\n",
    "exon_lengths_pr = (mouse_gr[mouse_gr.Feature == \"exon\"].df.assign(length=lambda df: df[\"End\"] - df[\"Start\"]).groupby(\"gene_id\")[\"length\"].sum().reset_index(name=\"total_exon_length\"))\n",
    "exon_lengths_sql = pd.read_sql_query(\"SELECT gene_id, SUM(End - Start) as total_exon_length FROM mouse WHERE Feature = 'exon' GROUP BY gene_id\", conn)\n",
    "exon_lengths_sql[\"total_exon_length\"] = exon_lengths_sql[\"total_exon_length\"].astype(int)\n",
    "exon_lengths_pr[\"total_exon_length\"] = exon_lengths_pr[\"total_exon_length\"].astype(int)\n",
    "if exon_lengths_sql.equals(exon_lengths_pr):\n",
    "    results_text = \"exon lengths are identical\"\n",
    "else:\n",
    "    results_text = \"exon lengths are not identical\"\n",
    "if OUTPUT_MODE == \"file\":\n",
    "    results_file.write(f\"{results_text}\\n\")\n",
    "else:\n",
    "    print(results_text)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Aggregate Query 3: Identify the chromosome with the highest number of transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using pyranges\n",
    "time = get_ipython().run_line_magic(\"timeit\",'-o transcript_counts_pr = mouse_gr[mouse_gr.Feature == \"transcript\"].df[\"Chromosome\"].value_counts().idxmax()')\n",
    "if OUTPUT_MODE == \"file\":\n",
    "    results_file.write(f\"Time taken to find chromosome with most transcripts using pyranges is {time}\\n\")\n",
    "else:\n",
    "    print(f\"Time taken to find chromosome with most transcripts using pyranges is {time}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# using sql\n",
    "time = get_ipython().run_line_magic(\"timeit\",'-o transcript_counts_sql = pd.read_sql_query(\"SELECT Chromosome, COUNT(*) as transcript_count FROM mouse WHERE Feature = \\'transcript\\' GROUP BY Chromosome ORDER BY transcript_count DESC LIMIT 1\", conn)')\n",
    "if OUTPUT_MODE == \"file\":\n",
    "    results_file.write(f\"Time taken to find chromosome with most transcripts using sql is {time}\\n\")\n",
    "else:\n",
    "    print(f\"Time taken to find chromosome with most transcripts using sql is {time}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "transcript counts are identical\n"
     ]
    }
   ],
   "source": [
    "# verify if transcript_counts_pr and transcript_counts_sql are identical\n",
    "transcript_counts_pr = mouse_gr[mouse_gr.Feature == \"transcript\"].df[\"Chromosome\"].value_counts().idxmax()\n",
    "transcript_counts_sql = pd.read_sql_query(\"SELECT Chromosome, COUNT(*) as transcript_count FROM mouse WHERE Feature = 'transcript' GROUP BY Chromosome ORDER BY transcript_count DESC LIMIT 1\", conn)\n",
    "if transcript_counts_pr == transcript_counts_sql[\"Chromosome\"].values[0]:\n",
    "    results_text = \"transcript counts are identical\"\n",
    "else:\n",
    "    results_text = \"transcript counts are not identical\"\n",
    "if OUTPUT_MODE == \"file\":\n",
    "    results_file.write(f\"{results_text}\\n\")\n",
    "else:\n",
    "    print(results_text)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Interval Arithmetic\n",
    "### Query 1: Merging Overlapping Exon Intervals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using pyranges\n",
    "time = get_ipython().run_line_magic(\"timeit\",'-o exon_intervals_pr = mouse_gr[mouse_gr.Feature == \"exon\"].merge(strand=True)')\n",
    "if OUTPUT_MODE == \"file\":\n",
    "    results_file.write(f\"Time taken to merge exons using pyranges is {time}\\n\")\n",
    "else:\n",
    "    print(f\"Time taken to merge exons using pyranges is {time}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# using sql\n",
    "def get_exon_intervals_sql(chrom_strand):\n",
    "    chrom, strand = chrom_strand\n",
    "    exon_intervals_sql = pd.read_sql_query(f\"SELECT Start, End FROM mouse WHERE Feature = 'exon' AND Chromosome = '{chrom}' AND Strand = '{strand}'\", conn)\n",
    "    exon_intervals_sql = pr.methods.merge._merge(exon_intervals_sql, chromosome=chrom, count=None, strand=strand)\n",
    "    return exon_intervals_sql\n",
    "\n",
    "def get_exon_intervals_sql_multi():\n",
    "    chrom_strand_tup = pd.read_sql_query(\"SELECT DISTINCT Chromosome, Strand FROM mouse\", conn)\n",
    "    chrom_strand_tup = list(zip(chrom_strand_tup[\"Chromosome\"], chrom_strand_tup[\"Strand\"]))\n",
    "    with Pool(8) as p:\n",
    "        exon_intervals_sql = p.map(get_exon_intervals_sql, chrom_strand_tup)\n",
    "    exon_intervals_sql = pd.concat(exon_intervals_sql).sort_values([\"Chromosome\", \"Strand\", \"Start\", \"End\"]).reset_index(drop=True)\n",
    "    return exon_intervals_sql"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "time = get_ipython().run_line_magic(\"timeit\",'-o exon_intervals_sql = get_exon_intervals_sql_multi()')\n",
    "if OUTPUT_MODE == \"file\":\n",
    "    results_file.write(f\"Time taken to merge exons using sql is {time}\\n\")\n",
    "else:\n",
    "    print(f\"Time taken to merge exons using sql is {time}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "exon intervals are identical\n"
     ]
    }
   ],
   "source": [
    "# check if exon_intervals_pr and exon_intervals_sql are identical\n",
    "exon_intervals_pr = mouse_gr[mouse_gr.Feature == \"exon\"].merge(strand=True).df.sort_values([\"Chromosome\", \"Strand\", \"Start\", \"End\"]).reset_index(drop=True)\n",
    "exon_intervals_sql = get_exon_intervals_sql_multi()\n",
    "if exon_intervals_pr.equals(exon_intervals_sql):\n",
    "    results_text = \"exon intervals are identical\"\n",
    "else:\n",
    "    results_text = \"exon intervals are not identical\"\n",
    "if OUTPUT_MODE == \"file\":\n",
    "    results_file.write(f\"{results_text}\\n\")\n",
    "else:\n",
    "    print(results_text)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Query 2: Finding Overlaps with a Specific Interval\n",
    "Find all gene features that overlap a given interval chr1:100000-200000 + strand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[17], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# Using pyranges\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m time \u001b[38;5;241m=\u001b[39m \u001b[43mget_ipython\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun_line_magic\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mtimeit\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m-o overlapping_genes_pr = mouse_gr[mouse_gr.Feature == \u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mgene\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m].overlap(pr.from_dict(\u001b[39;49m\u001b[38;5;124;43m{\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mChromosome\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m: [\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m1\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m], \u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mStart\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m: [100000], \u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mEnd\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m: [1000000], \u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mStrand\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m: [\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m+\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m]}), strandedness=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43msame\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m)\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m OUTPUT_MODE \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfile\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m      4\u001b[0m     results_file\u001b[38;5;241m.\u001b[39mwrite(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTime taken to find overlapping genes using pyranges is \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mtime\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/Desktop/code/genomics/.venv/lib/python3.12/site-packages/IPython/core/interactiveshell.py:2480\u001b[0m, in \u001b[0;36mInteractiveShell.run_line_magic\u001b[0;34m(self, magic_name, line, _stack_depth)\u001b[0m\n\u001b[1;32m   2478\u001b[0m     kwargs[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mlocal_ns\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mget_local_scope(stack_depth)\n\u001b[1;32m   2479\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbuiltin_trap:\n\u001b[0;32m-> 2480\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mfn\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2482\u001b[0m \u001b[38;5;66;03m# The code below prevents the output from being displayed\u001b[39;00m\n\u001b[1;32m   2483\u001b[0m \u001b[38;5;66;03m# when using magics with decorator @output_can_be_silenced\u001b[39;00m\n\u001b[1;32m   2484\u001b[0m \u001b[38;5;66;03m# when the last Python token in the expression is a ';'.\u001b[39;00m\n\u001b[1;32m   2485\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(fn, magic\u001b[38;5;241m.\u001b[39mMAGIC_OUTPUT_CAN_BE_SILENCED, \u001b[38;5;28;01mFalse\u001b[39;00m):\n",
      "File \u001b[0;32m~/Desktop/code/genomics/.venv/lib/python3.12/site-packages/IPython/core/magics/execution.py:1199\u001b[0m, in \u001b[0;36mExecutionMagics.timeit\u001b[0;34m(self, line, cell, local_ns)\u001b[0m\n\u001b[1;32m   1196\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m time_number \u001b[38;5;241m>\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0.2\u001b[39m:\n\u001b[1;32m   1197\u001b[0m             \u001b[38;5;28;01mbreak\u001b[39;00m\n\u001b[0;32m-> 1199\u001b[0m all_runs \u001b[38;5;241m=\u001b[39m \u001b[43mtimer\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrepeat\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrepeat\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mnumber\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1200\u001b[0m best \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mmin\u001b[39m(all_runs) \u001b[38;5;241m/\u001b[39m number\n\u001b[1;32m   1201\u001b[0m worst \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mmax\u001b[39m(all_runs) \u001b[38;5;241m/\u001b[39m number\n",
      "File \u001b[0;32m/usr/lib/python3.12/timeit.py:208\u001b[0m, in \u001b[0;36mTimer.repeat\u001b[0;34m(self, repeat, number)\u001b[0m\n\u001b[1;32m    206\u001b[0m r \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m    207\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(repeat):\n\u001b[0;32m--> 208\u001b[0m     t \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtimeit\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnumber\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    209\u001b[0m     r\u001b[38;5;241m.\u001b[39mappend(t)\n\u001b[1;32m    210\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m r\n",
      "File \u001b[0;32m~/Desktop/code/genomics/.venv/lib/python3.12/site-packages/IPython/core/magics/execution.py:173\u001b[0m, in \u001b[0;36mTimer.timeit\u001b[0;34m(self, number)\u001b[0m\n\u001b[1;32m    171\u001b[0m gc\u001b[38;5;241m.\u001b[39mdisable()\n\u001b[1;32m    172\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 173\u001b[0m     timing \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43minner\u001b[49m\u001b[43m(\u001b[49m\u001b[43mit\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtimer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    174\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m    175\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m gcold:\n",
      "File \u001b[0;32m<magic-timeit>:1\u001b[0m, in \u001b[0;36minner\u001b[0;34m(_it, _timer)\u001b[0m\n",
      "File \u001b[0;32m~/Desktop/code/genomics/.venv/lib/python3.12/site-packages/pandas/core/ops/common.py:76\u001b[0m, in \u001b[0;36m_unpack_zerodim_and_defer.<locals>.new_method\u001b[0;34m(self, other)\u001b[0m\n\u001b[1;32m     72\u001b[0m             \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mNotImplemented\u001b[39m\n\u001b[1;32m     74\u001b[0m other \u001b[38;5;241m=\u001b[39m item_from_zerodim(other)\n\u001b[0;32m---> 76\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mmethod\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mother\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Desktop/code/genomics/.venv/lib/python3.12/site-packages/pandas/core/arraylike.py:40\u001b[0m, in \u001b[0;36mOpsMixin.__eq__\u001b[0;34m(self, other)\u001b[0m\n\u001b[1;32m     38\u001b[0m \u001b[38;5;129m@unpack_zerodim_and_defer\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m__eq__\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     39\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21m__eq__\u001b[39m(\u001b[38;5;28mself\u001b[39m, other):\n\u001b[0;32m---> 40\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_cmp_method\u001b[49m\u001b[43m(\u001b[49m\u001b[43mother\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43moperator\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43meq\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Desktop/code/genomics/.venv/lib/python3.12/site-packages/pandas/core/series.py:6119\u001b[0m, in \u001b[0;36mSeries._cmp_method\u001b[0;34m(self, other, op)\u001b[0m\n\u001b[1;32m   6116\u001b[0m lvalues \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_values\n\u001b[1;32m   6117\u001b[0m rvalues \u001b[38;5;241m=\u001b[39m extract_array(other, extract_numpy\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, extract_range\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[0;32m-> 6119\u001b[0m res_values \u001b[38;5;241m=\u001b[39m \u001b[43mops\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcomparison_op\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlvalues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrvalues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mop\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   6121\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_construct_result(res_values, name\u001b[38;5;241m=\u001b[39mres_name)\n",
      "File \u001b[0;32m~/Desktop/code/genomics/.venv/lib/python3.12/site-packages/pandas/core/ops/array_ops.py:344\u001b[0m, in \u001b[0;36mcomparison_op\u001b[0;34m(left, right, op)\u001b[0m\n\u001b[1;32m    341\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m invalid_comparison(lvalues, rvalues, op)\n\u001b[1;32m    343\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m lvalues\u001b[38;5;241m.\u001b[39mdtype \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mobject\u001b[39m \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(rvalues, \u001b[38;5;28mstr\u001b[39m):\n\u001b[0;32m--> 344\u001b[0m     res_values \u001b[38;5;241m=\u001b[39m \u001b[43mcomp_method_OBJECT_ARRAY\u001b[49m\u001b[43m(\u001b[49m\u001b[43mop\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlvalues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrvalues\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    346\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    347\u001b[0m     res_values \u001b[38;5;241m=\u001b[39m _na_arithmetic_op(lvalues, rvalues, op, is_cmp\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n",
      "File \u001b[0;32m~/Desktop/code/genomics/.venv/lib/python3.12/site-packages/pandas/core/ops/array_ops.py:129\u001b[0m, in \u001b[0;36mcomp_method_OBJECT_ARRAY\u001b[0;34m(op, x, y)\u001b[0m\n\u001b[1;32m    127\u001b[0m     result \u001b[38;5;241m=\u001b[39m libops\u001b[38;5;241m.\u001b[39mvec_compare(x\u001b[38;5;241m.\u001b[39mravel(), y\u001b[38;5;241m.\u001b[39mravel(), op)\n\u001b[1;32m    128\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 129\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mlibops\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mscalar_compare\u001b[49m\u001b[43m(\u001b[49m\u001b[43mx\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mravel\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43my\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mop\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    130\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m result\u001b[38;5;241m.\u001b[39mreshape(x\u001b[38;5;241m.\u001b[39mshape)\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# Using pyranges\n",
    "time = get_ipython().run_line_magic(\"timeit\",'-o overlapping_genes_pr = mouse_gr[mouse_gr.Feature == \"gene\"].overlap(pr.from_dict({\"Chromosome\": [\"chr1\"], \"Start\": [4000000], \"End\": [5000000], \"Strand\": [\"+\"]}), strandedness=\"same\")')\n",
    "if OUTPUT_MODE == \"file\":\n",
    "    results_file.write(f\"Time taken to find overlapping genes using pyranges is {time}\\n\")\n",
    "else:\n",
    "    print(f\"Time taken to find overlapping genes using pyranges is {time}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# using sql\n",
    "other_genes_all = pd.DataFrame({\"Chromosome\": [\"chr1\"], \"Start\": [4000000], \"End\": [5000000], \"Strand\": [\"+\"]})\n",
    "def get_overlapping_genes_sql(chrom_strand):\n",
    "    chrom, strand = chrom_strand\n",
    "    self_genes_sql = pd.read_sql_query(f\"SELECT * FROM mouse WHERE Feature = 'gene' AND Chromosome = '{chrom}' AND Strand = '{strand}'\", conn)\n",
    "    other_genes = other_genes_all[(other_genes_all[\"Chromosome\"] == chrom) & (other_genes_all[\"Strand\"] == strand)]\n",
    "    overlapping_genes_sql = pr.methods.intersection._overlap(self_genes_sql, other_genes, how=\"first\")\n",
    "    return overlapping_genes_sql\n",
    "\n",
    "def get_overlapping_genes_sql_multi():\n",
    "    chrom_strand_tup = pd.read_sql_query(\"SELECT DISTINCT Chromosome, Strand FROM mouse\", conn)\n",
    "    chrom_strand_tup = list(zip(chrom_strand_tup[\"Chromosome\"], chrom_strand_tup[\"Strand\"]))\n",
    "    with Pool(8) as p:\n",
    "        overlapping_genes_sql = p.map(get_overlapping_genes_sql, chrom_strand_tup)\n",
    "    overlapping_genes_sql = pd.concat(overlapping_genes_sql) #.sort_values([\"Chromosome\", \"Strand\", \"Start\", \"End\"]).reset_index(drop=True)\n",
    "    return overlapping_genes_sql"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "time = get_ipython().run_line_magic(\"timeit\",'-o overlapping_genes_sql = get_overlapping_genes_sql_multi()')\n",
    "if OUTPUT_MODE == \"file\":\n",
    "    results_file.write(f\"Time taken to find overlapping genes using sql is {time}\\n\")\n",
    "else:\n",
    "    print(f\"Time taken to find overlapping genes using sql is {time}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "overlapping genes are identical\n"
     ]
    }
   ],
   "source": [
    "# check if overlapping_genes_pr and overlapping_genes_sql are identical\n",
    "overlapping_genes_pr = mouse_gr[mouse_gr.Feature == \"gene\"].overlap(pr.from_dict({\"Chromosome\": [\"chr1\"], \"Start\": [4000000], \"End\": [5000000], \"Strand\": [\"+\"]}), strandedness=\"same\")\n",
    "overlapping_genes_sql = get_overlapping_genes_sql_multi()\n",
    "overlapping_genes_sql = overlapping_genes_sql.sort_values([\"Chromosome\", \"Strand\", \"Start\", \"End\"]).reset_index(drop=True)\n",
    "overlapping_genes_pr = overlapping_genes_pr.df.sort_values([\"Chromosome\", \"Strand\", \"Start\", \"End\"]).reset_index(drop=True)\n",
    "diff = overlapping_genes_pr.compare(overlapping_genes_sql)\n",
    "if diff.empty:\n",
    "    results_text = \"overlapping genes are identical\"\n",
    "else:\n",
    "    results_text = \"overlapping genes are not identical\"\n",
    "if OUTPUT_MODE == \"file\":\n",
    "    results_file.write(f\"{results_text}\\n\")\n",
    "else:\n",
    "    print(results_text)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Query 3: Subtracting Intervals\n",
    "Remove a set of repetitive regions from the exon features.\n",
    "\n",
    "chr1:15M-16M,20M-21M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "other_cdf = pr.from_dict({\"Chromosome\": [\"chr1\", \"chr1\"],\"Start\": [15000000, 20000000],\"End\": [16000000, 21000000], \"Strand\": [\"+\", \"-\"]})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# using pyranges\n",
    "time = get_ipython().run_line_magic(\"timeit\",'-o subtracted_intervals_pr = mouse_gr[mouse_gr.Feature == \"exon\"].subtract(other_cdf, strandedness=\"same\")')\n",
    "if OUTPUT_MODE == \"file\":\n",
    "    results_file.write(f\"Time taken to subtract intervals using pyranges is {time}\\n\")\n",
    "else:\n",
    "    print(f\"Time taken to subtract intervals using pyranges is {time}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# using sql\n",
    "def get_subtracted_intervals_sql(chrom_strand):\n",
    "    chrom, strand = chrom_strand\n",
    "    self_genes_sql = pd.read_sql_query(f\"SELECT * FROM mouse WHERE Feature = 'exon' AND Chromosome = '{chrom}' AND Strand = '{strand}'\", conn)\n",
    "    other_cdf_clusters = other_cdf.merge(strand=True)\n",
    "    other_chrom_clusters = other_cdf_clusters[(other_cdf_clusters.Chromosome == chrom) & (other_cdf_clusters.Strand == strand)].df\n",
    "    # add __num__ column to self_genes_sql which counts how many intervals in self_genes_sql overlap with other_genes\n",
    "    self_genes_sql = pr.methods.coverage._number_overlapping(self_genes_sql, other_chrom_clusters, strandedness=\"same\", keep_nonoverlapping=True, overlap_col=\"__num__\")\n",
    "    subtracted_intervals_sql = pr.methods.subtraction._subtraction(self_genes_sql, other_chrom_clusters, strandedness=\"same\")\n",
    "    return subtracted_intervals_sql.drop(columns=[\"__num__\"])\n",
    "\n",
    "def get_subtracted_intervals_sql_multi():\n",
    "    chrom_strand_tup = pd.read_sql_query(\"SELECT DISTINCT Chromosome, Strand FROM mouse\", conn)\n",
    "    chrom_strand_tup = list(zip(chrom_strand_tup[\"Chromosome\"], chrom_strand_tup[\"Strand\"]))\n",
    "    with Pool(8) as p:\n",
    "        subtracted_intervals_sql = p.map(get_subtracted_intervals_sql, chrom_strand_tup)\n",
    "    subtracted_intervals_sql = pd.concat(subtracted_intervals_sql) #.sort_values([\"Chromosome\", \"Strand\", \"Start\", \"End\"]).reset_index(drop=True)\n",
    "    return subtracted_intervals_sql"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "time = get_ipython().run_line_magic(\"timeit\",'-o subtracted_intervals_sql = get_subtracted_intervals_sql_multi()')\n",
    "if OUTPUT_MODE == \"file\":\n",
    "    results_file.write(f\"Time taken to subtract intervals using sql is {time}\\n\")\n",
    "else:\n",
    "    print(f\"Time taken to subtract intervals using sql is {time}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "subtracted intervals are identical\n"
     ]
    }
   ],
   "source": [
    "# check if subtracted_intervals_pr and subtracted_intervals_sql are identical\n",
    "subtracted_intervals_pr = mouse_gr[mouse_gr.Feature == \"exon\"].subtract(other_cdf, strandedness=\"same\")\n",
    "subtracted_intervals_sql = get_subtracted_intervals_sql_multi()\n",
    "subtracted_intervals_sql.to_csv(\"subtracted_intervals_sql.csv\", index=False)\n",
    "subtracted_intervals_pr.df.to_csv(\"subtracted_intervals_pr.csv\", index=False)\n",
    "os.system(\"cat subtracted_intervals_sql.csv | sort > subtracted_intervals_sql_sorted.csv\")\n",
    "os.system(\"cat subtracted_intervals_pr.csv | sort > subtracted_intervals_pr_sorted.csv\")\n",
    "diff = os.popen(\"diff subtracted_intervals_sql_sorted.csv subtracted_intervals_pr_sorted.csv\").read()\n",
    "if diff == \"\":\n",
    "    results_text = \"subtracted intervals are identical\"\n",
    "else:\n",
    "    results_text = \"subtracted intervals are not identical:\" + diff\n",
    "if OUTPUT_MODE == \"file\":\n",
    "    results_file.write(f\"{results_text}\\n\")\n",
    "else:\n",
    "    print(results_text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results_file.close()\n",
    "conn.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
